Fix nested null value overrides

aeijdenberg opened the pull request:
Closes #5184 

**What this PR does / why we need it**:

This PR adds tests for setting nested values to `null` and ensure such values are actually deleted.

**Special notes for your reviewer**:

I've split this into 4 commits. They do the following:

1. Adds some code to the tests to make it easier to add nested null values tests.
2. Adds tests that fails.
3. Does a fairly minimal fix - this is enough to make the tests pass.
4. The last commit touches a bit more code, and attempts to simplify the coalescence more broadly by consolidating on to a single codepath, and being a bit more careful about which maps are mutated. Tests continue to pass - but I'm a little less confident about the full impact of this commit.

**If applicable**:

- [x] this PR contains unit tests

aeijdenberg committed:
Fix nested null value overrides

- Add ability to test for nested non-existent keys
- Add test cases for nested null values
- Minimalist fix for nested null key test cases
- Add missing metadata to integration test

Signed-off-by: Adam Eijdenberg <adam.eijdenberg@digital.gov.au>

cdenneen commented:
@aeijdenberg can you fix this by adding metadata to the integration tests? Really want to get this fix in place.

aeijdenberg commented:
@cdenneen - I've found the missing metadata, added it, and removed the nil check. Let me know if you'd like that added back in.

(oh, and also re-based on latest `master`)

cdenneen commented:
I’m okay with nil being valid “null” setting but that’s just me

cdenneen commented:
@aeijdenberg can you add label `bug` and `awaiting review`

cdenneen commented:
@aeijdebberg can you please add those labels?

bacongobbler commented:
Only admins can attach labels. I've done the needful

cdenneen commented:
Thanks @bacongobbler

scottrigby commented:
Hi I just saw this from #5184 👋 I'm late to the party

@cdenneen Quick reply to https://github.com/helm/helm/pull/5185#issuecomment-477104181: the reason I added [these 5 lines](https://github.com/helm/helm/blob/master/pkg/chartutil/values_test.go#L282-L286) in the test initially is that `null`, `Null`, `NULL`, and `~` are all valid when parsed, but different from empty string `""`. I think it's fine that these are already checked on these lines and don't need to be re-checked in the deeply nested test added by this PR 👍

@aeijdenberg re point 4 in the PR description, I'm also not fully sure about other possible impacts of this (second to last) commit: e69e8887cefb5b93e0e6dea7d82cb1fd819e1477. @bacongobbler I'd like another set of eyes on that. But, it does fix the 🐛 (the updated tests are great, and I also manually tested with a few different scenarios I'm aware of form community helm/charts) - everything seems to work 💯

scottrigby commented on pkg/chartutil/values.go:
```diff
@@ -326,25 +275,50 @@ func coalesceValues(c *chart.Chart, v map[string]interface{}) (map[string]interf
+			rv[key] = coalesceTables(dstTable, srcTable, chartName)
+		case srcIsTable && !dstIsTable:
+			log.Printf("Warning: Merging destination map for chart '%s'. Cannot overwrite table item '%s', with non table value: %v", chartName, key, val)
+			// despite message in warning, we appear to do exactly that, and do take the dst value
```
@aeijdenberg Yes this is the only thing I noticed here. Somehow, when nullifying a key this warning prints. Example from community helm/charts:
```console
$ helm install stable/drupal --set livenessProbe.httpGet=null --set livenessProbe.exec.command[0]='php -v' --debug --dry-run
...
2019/04/23 14:00:41 Warning: Merging destination map for chart 'drupal'. Cannot overwrite table item 'httpGet', with non table value: map[path:/user/login port:http]
```

I agree with your comment - this warning is confusing. It strangely seems to imply we're attempting to overwrite the original key with its value.

I wonder though, how this message ever triggers for a nullified key, since we `continue` above in that case?

@hickeyma since you wrote #4974 would you be willing to take a look at that side of these code changes?

hickeyma replied:
#4974 does not change any existing logic. It adds chart name to existing warning/error messages with some modifications to the message.

From looking at the code, this message is generated from [here](https://github.com/helm/helm/blob/master/pkg/chartutil/values.go#L336)
 
It is part of the if/else ladder:
```
if innerdst, ok := dst[key]; !ok {
	dst[key] = val
} else if istable(innerdst) {
	coalesceTables(innerdst.(map[string]interface{}), val.(map[string]interface{}), chartName)
} else {
	log.Printf("Warning: Merging destination map for chart '%s'. Cannot overwrite table item '%s', with non table value: %v", chartName, key, val)
}
continue
```
It seems to check that the destination map is valid yaml table and if is not, it prints the warning message. It continues after the ladder.

hickeyma replied:
Sorry, was looking at previous version of the file and not the updated one.

From my reading of the code (and not known the background to it), it would seem that you could have YAML returned from `dv, ok := dst[key]` which would not be null but would not match a valid YAML table and therefore be caught here: `case srcIsTable && !dstIsTable:` and so output the warning you mentioned.

scottrigby replied:
Never mind, somehow I must have made a mistake in debugging. I don't see the error now 😅 Resolving this one!

aeijdenberg commented:
Hi all, sorry to go quiet on this PR. Am back on deck now - let me know if any changes are desired.

scottrigby commented:
I resolved my last comment. This LGTM 👍 but would still like another set of eyes on the reorganizing commit I mentioned here https://github.com/helm/helm/pull/5185#issuecomment-485919391

thomastaylor312 commented:
Hey @aeijdenberg so that refactor commit looks pretty good, but I want to review it a little more in depth with @scottrigby to make sure it doesn't affect other things. Would you be willing to pop that commit off of this PR and open it as a separate PR? It would allow us to merge this fix and address the refactor in another PR.

Thanks for going through the refactoring effort with this code. We really appreciate it!

aeijdenberg commented:
@thomastaylor312 - done. I've rebased on latest `master`, squashed the commits (excluding the refactor), and have opened #5821 for the refactoring.

scottrigby reviewed (approved):
Thanks for splitting this. Lgtm

thomastaylor312 reviewed (approved):
Ok, sorry for the delay. I was talking with some of the other maintainers to make sure that deleting a key wouldn't create any side effects. This LGTM. Thanks for the good work!
